### PR TITLE
Allow using async callbacks in `add_next_tick_callback` and the like

### LINKED ISSUE #9658: [BUG] Unable to use async callbacks in `add_next_tick_callback` and the like
### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
Bokeh 2.0.0dev6

#### Description of expected behavior and the observed behavior
It should be possible to use async callbacks in `add_next_tick_callback`. However, such callbacks are never executed and result in `RuntimeWarning: coroutine 'f' was never awaited`.

#### Complete, minimal, self-contained example code that reproduces the issue
Should be used in an app for `bokeh.server.server.Server`:
```python
async def f():
    print('hi')

doc.add_next_tick_callback(f)
```

#### Stack traceback and/or browser JavaScript console output
```
/.../python3.7/asyncio/base_events.py:1760: RuntimeWarning: coroutine 'setup_theta_analysis.<locals>.f' was never awaited
Coroutine created at (most recent call last)
  File "/.../python3.7/site-packages/tornado/platform/asyncio.py", line 148, in start
    self.asyncio_loop.run_forever()
  File "/.../python3.7/asyncio/base_events.py", line 539, in run_forever
    self._run_once()
  File "/.../python3.7/asyncio/base_events.py", line 1767, in _run_once
    handle._run()
  File "/.../python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/.../python3.7/site-packages/bokeh/server/session.py", line 65, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/.../python3.7/site-packages/bokeh/server/session.py", line 181, in with_document_locked
    return func(*args, **kwargs)
  File "/.../python3.7/site-packages/bokeh/document/document.py", line 1160, in wrapper
    return doc._with_self_as_curdoc(invoke)
  File "/.../python3.7/site-packages/bokeh/document/document.py", line 1146, in _with_self_as_curdoc
    return f()
  File "/.../python3.7/site-packages/bokeh/document/document.py", line 1159, in invoke
    return f(*args, **kwargs)
  File "/.../python3.7/site-packages/bokeh/document/document.py", line 949, in remove_then_invoke
    return callback(*args, **kwargs)
  handle = self._ready.popleft()
```

### PR DESCRIPTION
**Pull request by p-himik**
- [x] issues: fixes #9658
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

File: bokeh/server/session.py [hunk 1]
```diff
@@ -63,6 +64,10 @@ def _needs_document_lock(func):
                 self._pending_writes = []
                 try:
                     result = func(self, *args, **kwargs)
+                    if inspect.iscoroutine(result):
```
Comment (bryevdv) [inline bokeh/server/session.py]: I think we want `isawaitable` here (a bit more precise)

https://docs.python.org/3/library/inspect.html#inspect.isawaitable

**Comment by bryevdv** [comment]
Is there a simple way to add a test so we make sure not to accidentally break this support in the future?

**Comment by p-himik** [comment]
I think so. I'll look into in tomorrow.

**Comment by p-himik** [comment]
Also, I have no idea what that test failure is about. Does it make sense to you?

**Comment by bryevdv** [comment]
Pretty sure it's just a flaky test, I've resrtarted

File: tests/unit/bokeh/server/test_contexts.py [hunk 2]
```diff
@@ -140,14 +134,32 @@ class FakeRequest(object):
...
-        class FakeRequest(object):
-            arguments = dict(foo=10)
-        req = FakeRequest()
-        s = await c.create_session_if_needed("foo", request=req)
+        s = await c.create_session_if_needed("foo")
         session = c.get_session("foo")
         assert session == s
         assert c._session_contexts[session.id].logout_url == "/logout"
 
+    @pytest.mark.asyncio
```
Comment (bryevdv) [inline tests/unit/bokeh/server/test_contexts.py]: will need to rebase on master and remove the decorator now after https://github.com/bokeh/bokeh/pull/9662

**Pull request merged**